### ISSUE TITLE
[used-before-assignment] Functions defined in type-checking guard not flagged when used later

### ISSUE THREAD

**Issue by jacobtylerwalls** [pylint-dev/pylint#10028]
In #9990, we [discovered](https://github.com/pylint-dev/pylint/pull/9990#issuecomment-2411463617) that extending the fix for false negatives for used-before-assignment for objects guarded behind type-checking blocks to include functions would cause a handful of false positives in the primer. So we backed out that change and opened this issue to investigate.

@zenlyj graciously offered to find time to look into it.

**Comment by zenlyj** [comment]
will also look at this `unused-import` false negative that @nickdrozd found.

```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from math import pi  # unused-import false negative
```
